From e629248a3fb8aced9246e642193cddb2b7abd3fa Mon Sep 17 00:00:00 2001
From: Zhang Ning <ning.a.zhang@intel.com>
Date: Mon, 19 Sep 2016 15:14:40 +0800
Subject: [PATCH 4260/4301] rootfs: use async to populate rootfs

commit 2e74eeec03a959bef7df727ca9f03d42aaa0b47c from
git://git@github.com:01org/linux-intel-4.9.git

this patch is originally from:
https://lists.ubuntu.com/archives/kernel-team/2009-December/007544.html
with modify to fit new kernel, and less build warning.

it can save 160ms on boot, and independs on bootimage size.

Signed-off-by: Zhang Ning <ning.a.zhang@intel.com>
---
 include/linux/init.h |    2 ++
 init/initramfs.c     |   12 +++++++++++-
 init/main.c          |    7 +++++++
 3 files changed, 20 insertions(+), 1 deletions(-)

diff --git a/include/linux/init.h b/include/linux/init.h
index 6935d02..8804a88 100644
--- a/include/linux/init.h
+++ b/include/linux/init.h
@@ -227,6 +227,8 @@ extern bool initcall_debug;
 	static initcall_t __initcall_##fn \
 	__used __section(.security_initcall.init) = fn
 
+extern struct async_domain populate_rootfs_domain;
+
 struct obs_kernel_param {
 	const char *str;
 	int (*setup_func)(char *);
diff --git a/init/initramfs.c b/init/initramfs.c
index b32ad7d..a15084b 100644
--- a/init/initramfs.c
+++ b/init/initramfs.c
@@ -18,6 +18,7 @@
 #include <linux/dirent.h>
 #include <linux/syscalls.h>
 #include <linux/utime.h>
+#include <linux/async.h>
 
 static ssize_t __init xwrite(int fd, const char *p, size_t count)
 {
@@ -605,7 +606,9 @@ static void __init clean_rootfs(void)
 }
 #endif
 
-static int __init populate_rootfs(void)
+ASYNC_DOMAIN(populate_rootfs_domain);
+
+static void __init async_populate_rootfs(void *data, async_cookie_t cookie)
 {
 	char *err = unpack_to_rootfs(__initramfs_start, __initramfs_size);
 	if (err)
@@ -653,6 +656,13 @@ static int __init populate_rootfs(void)
 		 */
 		load_default_modules();
 	}
+	return;
+}
+
+static int __init populate_rootfs(void)
+{
+	async_schedule_domain(async_populate_rootfs, NULL, &populate_rootfs_domain);
 	return 0;
 }
+
 rootfs_initcall(populate_rootfs);
diff --git a/init/main.c b/init/main.c
index 2c14b44..d028fac 100644
--- a/init/main.c
+++ b/init/main.c
@@ -1029,6 +1029,13 @@ static noinline void __init kernel_init_freeable(void)
 	(void) sys_dup(0);
 	(void) sys_dup(0);
 	/*
+	 * We need to ensure that the filesystem is ready by this point,
+	 *  wait for async_populate_rootfs to complete.
+	 */
+#ifdef CONFIG_BLK_DEV_INITRD 
+	async_synchronize_full_domain(&populate_rootfs_domain);
+#endif
+	/*
 	 * check if there is an early userspace init.  If yes, let it do all
 	 * the work
 	 */
-- 
1.7.5.4

